### PR TITLE
feat: add contribution guidelines

### PR DESCRIPTION
**Pull request by SuperFluffy**
## Summary
Add contribution guidelines summarizing how code is written at Astria (at least targetting the monorepo).

## Background
With the Astria getting new team members and the project growing (potentially inviting external contributors), we need more formal documents around how to contribute and what ideal PRs look like.

## Changes
* Add a `CONTRIBUTING.md` to the root of the repo.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+Before an author submits a patch to the Astria monorepo at https://github.com/astriaorg/astria,
+they should make sure to take into account the guidelines laid out in this document.
+
+## How Pull Requests are written at Astria
+
+The following points represent an ideal scenario that authors of patches should try to
+follow when submitting them in the form of GitHub pull requests (PRs):
+
+0. Be kind and respectful or your readers. Code is written once and read many times. Be mindful
```
Comment (noot) [inline CONTRIBUTING.md:11]:
```suggestion
0. Be kind and respectful of your readers. Code is written once and read many times. Be mindful
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+5. To point 3, separate refactoring and feature development into separate PRs. If a feature
+   requires the codebase be refactored, refactor first, implement the feature in a follow-up PR.
+   The same applies to bugfixes.
+6. When moving around code in refactors (for example, moving a test module into a separate file),
+   do not also change item names or add new items. Remember that GitHub diffs do not help a reviewer
+   track which code is new, deleted, or just moved.
+7. PR title and summary should give a concise and accurate representation of what the patch
+   wants to accomplish. They should not make assumptions about a reader knowing implementation details.
+   Ideally, put these in doc comments that live next to the code.
+8. If a code change does not help accomplish the patche's goal described in 7, it should be moved to
```
Comment (noot) [inline CONTRIBUTING.md:26]:
```suggestion
8. If a code change does not help accomplish the patch's goal described in 7, it should be moved to
```

**Review by noot — COMMENTED** [review]
nice, this is good to have. just flagging the markdown linter

also, should we add things like "make sure to run fmt, clippy" etc? do we want to ensure every PR corresponds to an issue that's already been accepted?

**Comment by SuperFluffy** [comment]
> nice, this is good to have. just flagging the markdown linter
> 
> also, should we add things like "make sure to run fmt, clippy" etc? do we want to ensure every PR corresponds to an issue that's already been accepted?

We could add this to the doc, but I wanted to get the "spirit" of writing PRs in here.

Although clippy and fmt will be flagged by CI anyways, so I think that's not really necessary.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+7. PR title and summary should give a concise and accurate representation of
+   what the patch wants to accomplish. They should not make assumptions about a
+   reader knowing implementation details. Ideally, put these in doc comments
+   that live next to the code.
+8. If a code change does not help accomplish the patch's goal described in 7,
+   it should be moved to another patch.
+9. Code should not be surprising - it should be idiomatic and fit in with the
+   rest of the code surrounding it. If a code changes is still necessary, flag
+   it for the reader using an inline comment, or adding a github review comment
+   pulling attention to it.
```
Comment (Fraser999) [inline CONTRIBUTING.md:36]: I'm not quite following the last sentence.  Do you mean if further changes are still necessary, the author should flag these?

Seems like that could be a separate point on its own, and be more general, i.e. the author should flag any points of special interest that way to help out the reviewer?  (I really liked that approach when I first saw it here btw.)

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+   or just moved.
+7. PR title and summary should give a concise and accurate representation of
+   what the patch wants to accomplish. They should not make assumptions about a
+   reader knowing implementation details. Ideally, put these in doc comments
+   that live next to the code.
+8. If a code change does not help accomplish the patch's goal described in 7,
+   it should be moved to another patch.
+9. Code should not be surprising - it should be idiomatic and fit in with the
+   rest of the code surrounding it. If a code changes is still necessary, flag
+   it for the reader using an inline comment, or adding a github review comment
```
Comment (Fraser999) [inline CONTRIBUTING.md:35]:
```suggestion
   it for the reader using an inline comment, or adding a GitHub review comment
```

**Review by Fraser999 — COMMENTED** [review]
Looks good to me.  The spirit seems to be keep the codebase and reviewers sane :) 

Do we also have recommendations on commits in a PR?  I know some folks prefer a single commit per PR, while at the other end of the spectrum some prefer commits to be tiny, with many commits in a single PR.  There's also sometimes a preference for every commit to be able to be compiled (i.e. no "work in progress" commits) to make using e.g. git bisect more useful.

Do we have a policy/recommendation of disallowing force-pushing to an open PR?  If it's allowed, do we have a preference of addressing review comments by force-pushing fixup commits or just pushing new commits with the fixes?

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+   that GitHub diffs do not help a reviewer track which code is new, deleted,
+   or just moved.
+7. PR title and summary should give a concise and accurate representation of
+   what the patch wants to accomplish. They should not make assumptions about a
+   reader knowing implementation details. Ideally, put these in doc comments
+   that live next to the code.
+8. If a code change does not help accomplish the patch's goal described in 7,
+   it should be moved to another patch.
+9. Code should not be surprising - it should be idiomatic and fit in with the
+   rest of the code surrounding it. If a code changes is still necessary, flag
```
Comment (joroshiba) [inline CONTRIBUTING.md:34]:
```suggestion
   rest of the code surrounding it. If a code change is still necessary, flag
```

**Review by noot — APPROVED** [review]
looks good! I'd prefer to add fmt/lint info in here somewhere too, even if it's caught by CI, someone should still run it locally before + newcomers are asking about it in the chat

**Comment by joroshiba** [comment]
> Looks good to me. The spirit seems to be keep the codebase and reviewers sane :)
> 
> Do we also have recommendations on commits in a PR? I know some folks prefer a single commit per PR, while at the other end of the spectrum some prefer commits to be tiny, with many commits in a single PR. There's also sometimes a preference for every commit to be able to be compiled (i.e. no "work in progress" commits) to make using e.g. git bisect more useful.
> 
> Do we have a policy/recommendation of disallowing force-pushing to an open PR? If it's allowed, do we have a preference of addressing review comments by force-pushing fixup commits or just pushing new commits with the fixes?

We don't have any recommendations here because I don't think we will find unity. We've had enough bike shedding of merge vs rebase workflows, haha. My preference is generally for a one commit and rebase workflow (akin to using mercurial), but github doesn't work great with this.

Generally with GitHub my preference is for no force push after PR opened, and commits on top in order to enable seeing changes since last review. I wish github was just better with managing snapshots though haha.

**Comment by joroshiba** [comment]
Mega old mostly agreed upon PR, we should get tidied up and merged

**Comment by joroshiba** [comment]
This PR is stale because it has been open 45 days with no activity. Remove stale label or this PR will be
closed in 7 days.

**Comment by joroshiba** [comment]
This PR was closed because it has been stale.

**Pull request closed without merging**